Rewrite `ref` example using custom modifier (#1509)
ember-learn/guides-source: guides/release/components/template-lifecycle-dom-and-modifiers.md

<button type="button">Pause</button>
```

How should we connect clicking the "Play" and "Pause" to calling the audio tag's `play` and `pause` methods?

While we *could* manage these DOM interactions in the component class (for example, by using `{{did-render}}`), we're better off using a modifier here. It lets us cleanly separate our concerns: the component manages the *state*, and the modifier manages *interactions with the DOM*.

There are three reasons to reach for modifiers for DOM element interactions:

> 1. A component, by itself, doesn't have direct access to DOM elements. We have to render the page, push an element back up into the component class, and only then can we safely refer to that element. This can sometimes require us to render the component's HTML twice in order for things to start working. Modifiers let us avoid this possible performance issue.

Reviewer: Thanks for clarifying what you meant by "wire it up". 💯

Author: Thanks for catching it as a point of confusion!

2. By keeping state in the component and handling DOM method calls in a modifier, we can use autotracking and stick to 1-way data flow in the component. Further, we could change the component's own design later *without* having to change how we interact with the DOM element.

3. The code for calling the audio element's `play` and `pause` can be reused. It isn't tied to this particular audio component. It can be tested independently, too!

> Now that we see *why* to use a modifier for our audio component, let's walk through *how* to create one. We will start with the component (to manage the state) and then implement the modifier (the manage the DOM).

Reviewer: ```suggestion
Now that we see *why* we want to use a modifier for our audio component, let's walk through *how* to create one. We will start with the component (to manage the state) and then implement the modifier (the manage the DOM).
```

<button type="button">Pause</button>
```

> It might be tempting to try to handle this by managing the whole set of interactions at the component level. But there are three reasons we might want to tackle this using a modifier instead:

Reviewer: Initially, I found the 3 reasons to be confusing (a bit too abstract). It wasn't until I reached the end of this subsection, where you wrote the tagline,

```
The component manages the state. The modifier manages the DOM.
```

that I understood what you wanted readers to do.

I think it'd be good to include these lines in the beginning, then reiterate them at the end.

```suggestion
It might be tempting to define all interactions in the component class. However, as we will see soon, the key to a better solution is to separate concerns. Let the component manage the state and the modifier manage the DOM.

In general, there are 3 reasons why we might want to introduce a modifier:
```

Author: Good callout! Committed that suggestion directly.

Author: As I'm mulling, I'm going to push a slight tweak here, but the overall direction I'm still very much 👍 on.

> 1. Components don't have access to DOM elements directly. We'd have to render the page, push an element back up into the component class, and *then* wire it up. This would come with performance costs, because we would have to render twice before anything could work. It would also 

Reviewer 2: This trails off oddly:

> It would also

Author: Heh, I turned this into the next bullet point and forgot to remove it. Thank you!

> 1. Components don't have access to DOM elements directly. We'd have to render the page, push an element back up into the component class, and *then* wire it up. This would come with performance costs, because we would have to render twice before anything could work.
> 2. We can't really use autotracking or 1-way data flow that way. If we wanted the caller of our component to be able to pass an argument in and have the component respond appropriately, we'd have to use something like `{{did-update}}`.
> 3. None of the logic we built to manage playing and pausing would be reusable! It would be specific to just this component.

Reviewer: ```suggestion
1. Components don't have direct access to DOM elements. We have to render the page, push an element into the component class, then wire it up. These steps can cause a performance issue because we have to render the page twice in order for things to start working.
2. We can't use autotracking and one-way data flow. If we want the caller of a component to be able to pass an argument and have the component respond appropriately, we will have to use a solution like `{{did-update}}`.
3. The code for managing play and pause is not reusable. It is tied to the audio component.
```

Author: This is helpful! I went with some slightly different wording and structure here, while trying to get at these goals; see 414d25ee.

> Using a custom modifier, we can avoid all of these issues!
> 
> First, let's add event handlers to the `Play` and `Pause` buttons:

Reviewer: ```suggestion
With these reasons in mind, let's look at how to create a modifier for our audio component. We will arrive at the final solution in incremental steps.

First, we add event handlers to the `Play` and `Pause` buttons:
```

Author: Took this idea and tweaked it a bit; see 414d25ee.

  play() {
    // TODO
  }

  @action
  pause() {
    // TODO
  }
}
```

> From the component's point of view, all it needs to do is track whether it should be playing or not, since we're going to manage the DOM element with the modifier. So we can just use `@tracked` and a normal property here.

Reviewer: Let's reiterate our point (separate concerns) for the 2nd time. 🙂

```suggestion
Recall that our modifier will manage the DOM (i.e. calling the audio element's `play` or `pause` method). All the component needs to do is to track if the audio is playing.
```

```js {data-filename="app/components/audio-player.js" data-diff="+2,+6,+7,-10,+11,-16,+17"}
import Component from "@glimmer/component";
import { tracked } from "@glimmer/tracking";
import { action } from "@ember/object";

export default class AudioPlayerComponent extends Component {
  @tracked isPlaying = false;

  @action
  play() {
    // TODO
    this.isPlaying = true;
  }

  @action
  pause() {
    // TODO
    this.isPlaying = false;
  }
}
```

> Now the component is managing the state of the player, but we still need to translate that into the DOM method calls. So let's create a modifier!

Reviewer: ```suggestion
Notice that the component manages the state of the audio player. We still need to translate the state into the appropriate DOM method call. Let's create the modifier now!
```

Author: Did a slight variant on this in b1da6ccd!

```bash
ember install ember-modifier
ember generate modifier play-when
```

> In this modifier, we'll take an argument that specifies whether we should be playing or paused.

Reviewer: ```suggestion
The modifier takes 1 argument, a Boolean that specifies if we should call the element's `play` or `pause` method.
```

Author: Much clearer!

> In this modifier, we'll simply take an argument that specifies whether we should be playing or paused.

Reviewer 2: I tend to prefer to avoid "simply" and "just" as a bit condescending, maybe its fine without?

```suggestion
In this modifier, we'll take an argument that specifies whether we should be playing or paused.
```

Author: Fixed via a force push. Thank you!

```js {data-filename="app/modifiers/play-when.js"}
import { modifier } from "ember-modifier";

export default modifier((element, [isPlaying]) => {
  if (isPlaying) {
    element.play();
  } else {
    element.pause();
  }
});
```

> We can now attach the modifier to the `audio` element:

Reviewer: ```suggestion
Last but not least, we attach the modifier to the `audio` element:
```

```handlebars {data-filename="app/components/audio-player.hbs" data-diff="-1,+2"}
<audio src={{@srcURL}} />
<audio src={{@srcURL}} {{play-when this.isPlaying}} />

> <button type="button" {{on "click" this.play}}>Play</button>

Reviewer 3: wondering, is it real use-case? (looks like we could remove `play-when` modifier and handle play logic in actions at all)

Reviewer 3: usecase from our project:

```js
export default modifier(function scrollTop(element, [enabled]) {
  if (enabled) {
    element.scrollIntoView({behavior: "smooth", block: "start", inline: "nearest"});
  }
});
```

Reviewer 3: or we could add more logic to `play-when`, to toggle `isPlaying` once audio will be finished (to get it close to real-life)

Author: A couple notes here—

1. This is in fact a real-world use case (as well as the example used originally here). The implementation I put up is *very* similar to one we designed for use in our own app.
2. My take on maintainability is a little different, I think: maintaining a very simple modifier and its tests is actually a big *win*, because it only has one thing it does and can focus on doing it well.

<button type="button" {{on "click" this.pause}}>Pause</button>
```

> With that, we're done! Notice that we've separated the concerns of *managing state* from *managing DOM interactions*. The component manages the state. The modifier manages the DOM. And this modifier could be reused for *any* `HTMLMediaElement`, including other `audio` or `video` element in our code.

Reviewer: ```suggestion
With that, we can now click the buttons to play and pause the audio!

In summary, when you want to allow elements in a component to communicate, see if you can separate the concerns of *managing state* and *managing DOM interactions*. The component can manage the state, while a modifier can manage the DOM.

The modifier that we made for the audio player component can be reused on *any* element that implements `play` and `pause` methods. In particular, we can reuse the modifier on an `HTMLMediaElement`, which includes `audio` and `video` elements.
```

Author: I *really* liked this; it makes the takeaway much clearer. Thank you!

}
```

> That's it for the component: we're translating the user's interactions into *state*. Now we need to build a modifier to translate the state into the appropriate DOM method calls!

Reviewer: I like these sentences. They reminded me of composing functions in math.

Author: Don’t tell anyone, but I’m kind of obsessed with composition in the mathematical sense as a key ingredient of software design. 🤓
